### ISSUE TITLE
db_trans bug

### ISSUE THREAD

**Issue by PeterChaplin** [qcode-software/qcode-tcl#303]
Example code:
```tcl
db_trans {
    ::try {
        db_trans {
            error "Something went wrong"
        }
    } on error {msg options} {
        # Error handler
    }
}
db_trans {
    db_dml {
        create table test
        (name text)
    }
    error "Something else went wrong"
}
```
After running this script, table `test` should not exist in the database, but it does.
Running `set ::db_trans_level(DEFAULT)` returns `-1`

**Comment by PeterChaplin** [comment]
Is there a reason we don't use postgreql SAVEPOINTs to handle nested transactions?

**Comment by PeterChaplin** [comment]
What is the desired behaviour of:
```tcl
db_dml {
    create table test
    (name text)
}

proc api_proc_1 {} {
    db_trans {
        db_dml {
            insert into test (name) values ('Bob')
        }
        error "Something went wrong"
    }
}

db_trans {
    ::try {
        api_proc_1
    } on error {msg options} {
    }
    db_dml {
        insert into test (name) values ('John')
    }
}
```
Should the insert of 'John' be committed, or rolled back?

**Comment by davidqc** [comment]
As an impartial observer I'd say the presence of the ```::try``` clause means the developer is explicitly saying they don't want errors inside ```api_proc_1``` to affect the current proc (except via whatever is in the ```on error``` clause).

So if I wrote this I'd expect John to be committed.

**Comment by davidqc** [comment]
```SAVEPOINT``` didn't exist in postgresql <9.1 so that's one reason it isn't used (there might be others)

**Comment by PeterChaplin** [comment]
I'm not sure there is a way to commit "John" and not "Bob" without using `SAVEPOINT`.

**Comment by davidqc** [comment]
(When I say I'd "expect" John to be committed, I really mean I'd "want" John to be committed)

**Comment by PeterChaplin** [comment]
The options I can think of would be

1. Use `SAVEPOINT`, which introduces a dependency on postgresql 9.1
2. Rollback everything if an error occurs inside a nested db_trans, even if the error is caught and handled between db_trans nesting layers.
3. As option 2, but also throw an error in the outer transaction
4. Use `SAVEPOINT` if it's available, otherwise as option 3

**Comment by PeterChaplin** [comment]
At present, if we run
```tcl
db_dml {
    create table test
    (name text)
}

proc api_proc_1 {} {
    db_trans {
        db_dml {
            insert into test (name) values ('Alice')
        }
        error "Something went wrong"
    }
}

db_trans {
    ::try {
        api_proc_1
    } on error {msg options} {
    }
    db_dml {
        insert into test (name) values ('Sally')
    }
    error "Something else went wrong"
}
```
"Sally" would be committed, and the global transaction level set to `-1`.

**Comment by daniel-qcode** [comment]
I think ```SAVEPOINT```s are the way to go, implementing option 3 as a fallback may not be necessary ... we could maybe jump to major release qcode-tcl-7.0.0 and stipulate that it requires postgresql version 9.1 and above.

@bvw what do you think.

**Comment by PeterChaplin** [comment]
Suggested changes:
https://github.com/qcode-software/qcode-tcl/compare/db_trans-bug
or
https://github.com/qcode-software/qcode-tcl/compare/db_trans-savepoints

**Comment by bvw** [comment]
Happy to go with `SAVEPOINT`s which I agree would be more consistent with expectations.
Are there any performance implications of using `SAVEPOINT` ?

On the example code a blanket `::try` may unwisely mask errors.

Style guide > Patterns
1. Try to limit the scope of exception handling. Don't catch all exceptions `catch {some code that might blow up}`

https://github.com/qcode-software/qcode/blob/master/wiki/Qcode-Style-Guide.md

So for the example rather than 
```tcl
db_trans {
    ::try {
        api_proc_1
    } on error {msg options} {
    }
    db_dml {
        insert into test (name) values ('Sally')
    }
    error "Something else went wrong"
}
```
use
```tcl
db_trans {
    api_proc_1
    db_dml {
        insert into test (name) values ('Sally')
    }
    error "Something else went wrong"
}
```

If we want to `try` to do something that might fail but not impede the rest of the transaction then we should design it to
1. Set a flag in the database which we know will succeed.
2. Schedule a task to independently and perhaps repeatedly attempt the required action against this object and unset the flag.

```tcl
db_trans {
    set_flag
    db_dml {
        insert into test (name) values ('Sally')
    }
    if { ! ok } { error "Something else went wrong" }
}
```

In scheduled task
```tcl
foreach thingy with flag set  {
    ::try {
            send_email
            unset flag
     } trap {SMTP ...} {
    ....
      }
}
```

**Comment by PeterChaplin** [comment]
Yeah, I was just trying to keep the code examples to the bare minimum needed to illustrate the problem, rather than working through the specifics of how/when/why a `try` would be used.

The specific case where this came up was in setting up test suites. I was wrapping each test case in a `db_trans`, with a `db_dml "ROLLBACK"` at the end, to try and ensure no changes were committed to the database. The `tcltest::test` proc catches and handles errors, though, so that if one test case fails with an error, it simply gets reported as a failure, and the test suite moves on to the next test.

**Comment by PeterChaplin** [comment]
Regarding performance, we could try running some tests, but I would not expect there to be any significant performance impact. If nothing else, we would only use savepoints for nested `db_trans`, which don't occur very often. I'm not actually sure how we would go about testing this.

**Comment by davidqc** [comment]
BTW I was misreading the Postgresql docs when I said SAVEPOINTS didn't exist before 9.1... they did, so no version dependency would be introduced by using them.

**Comment by daniel-qcode** [comment]
I prefer your suggested change to use postgresql SAVEPONTS:
https://github.com/qcode-software/qcode-tcl/compare/db_trans-savepoints

Could you create a pull request and add a few more tcltests:
   * tests for nested db_trans that all succeed
   * tests for sibling db_trans 
   * any others you can think of

I think that MLA's and TLC's DEV, TESTING and LIVE  databases are all postgresql version 9.1 or above.

**Comment by PeterChaplin** [comment]
https://github.com/qcode-software/qcode-tcl/pull/305
